### ISSUE TITLE
`prefer-const` not triggered (regression from 2.4.0)

### ISSUE THREAD

**Issue by fabienjuif** [eslint/eslint#5837]
**What version of ESLint are you using?**
2.7.0

**What parser (default, Babel-ESLint, etc.) are you using?**
>  "babel-eslint": "~6.0.0"

**Please show your full configuration:**
```
{
  "parser": "babel-eslint",
  "extends": "airbnb",
  "rules": {
      "semi": [2, "never"],
      "arrow-body-style": 0
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```
import React, { PropTypes } from 'react'

const innerStyle = {
  color: '#343434',
  color2: 'white',
}

const ToolBar = ({ style, draw }) => {
  let dynStyle = {
    ...innerStyle,
    ...style,
  }

  dynStyle.color = draw ? dynStyle.color2 : dynStyle.color

  return (
    <div style={dynStyle} />
  )
}

ToolBar.propTypes = {
  style: PropTypes.object,
  draw: PropTypes.bool,
}

export default ToolBar
```

**What did you expect to happen?**
`prefer-const` should trigger and ask me to change `let dynStyle` to `const dynStyle`

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint don't trigger the `prefer-const` rule.

---

Note I tried these versions :
 - `~2.4.0` : **Works** 
 - `~2.5.0` : **Doesn't** work
 - `~2.6.0` : **Doesn't** work
 - `~2.7.0` : **Doesn't** work

---

Commits that are related to `prefer-const` (maybe I miss some) :
 - `2.6`: 9b73ffd Update: destructuring option of prefer-const rule (fixes #5594) (Toru Nagashima)
 - `2.5`: 9a22625 Fix: prefer-const false positive at non-blocked if (fixes #5610) (Toru Nagashima)
 - `2.5`: 87d74b2 Fix: prefer-const got to not change scopes (refs #5284) (Toru Nagashima)

**Comment by ilyavolodin** [comment]
@mysticatea Could you take a look please?

**Comment by nzakas** [comment]
@fabienjuif 
It looks like you're using `babel-eslint` as your parser. Can you try using the default parser instead? You can just delete the `parser` from your config file to switch to the default parser. Make sure you configure the [parser options](http://eslint.org/docs/user-guide/configuring#specifying-parser-options) as well.

**Comment by fabienjuif** [comment]
**Config file :**
```
{
  "parserOptions": {
    "ecmaVersion": 7,
    "sourceType": "module",
    "ecmaFeatures": {
        "jsx": true,
        "experimentalObjectRestSpread": true
    }
  },
  "extends": "airbnb",
  "rules": {
      "semi": [2, "never"],
      "arrow-body-style": 0
  }
}
```

**Code :** Is the same
**How do I run it  ? :** ./node_modules/eslint/bin/eslint.js _[...]_/ToolBar.js
**Result :** Blank.

**Comment by mysticatea** [comment]
Thank you for this issue.

Hm, I cannot reproduce this.
`prefer-const` is warning `dynStyle`.

```
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> cat test.js
import React, { PropTypes } from 'react'

const innerStyle = {
  color: '#343434',
  color2: 'white',
}

const ToolBar = ({ style, draw }) => {
  let dynStyle = {
    ...innerStyle,
    ...style,
  }

  dynStyle.color = draw ? dynStyle.color2 : dynStyle.color

  return (
    <div style={dynStyle} />
  )
}

ToolBar.propTypes = {
  style: PropTypes.object,
  draw: PropTypes.bool,
}

export default ToolBar

C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint --version
v2.7.0
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint test.js --rule prefer-const:2 --parser-options "{ecmaVersion:7,sourceType:'module',ecmaFeatures:{jsx:true,experimentalObjectRestSpread:true}}" --no-ignore --no-eslintrc

C:\Users\t-nagashima.AD\Documents\GitHub\eslint\test.js
  9:7  error  'dynStyle' is never reassigned, use 'const' instead  prefer-const

✖ 1 problem (1 error, 0 warnings)

**Comment by stevoland** [comment]
I had this problem and could reproduce the test case but it was fixed when I removed `react/jsx-uses-vars` https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-uses-vars.js from my config

**Comment by stevoland** [comment]
That plugin flags `eslintUsed`: https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/util/variable.js

**Comment by nzakas** [comment]
@stevoland interesting! @fabienjuif can you check that?

**Comment by lukeapage** [comment]
yes thats right - I raised this on the react plugin - https://github.com/yannickcr/eslint-plugin-react/issues/716
I've created a test repo https://github.com/lukeapage/eslint-react-bug-prefer-const-716

after some investigation, as @stevoland suggested, eslint-plugin-react marks the variable with eslintUsed which then causes prefer-const to ignore the variable because of this line - https://github.com/eslint/eslint/blob/421e4bf933098c2615747c5bc2cef6b549f5f6c5/lib/rules/prefer-const.js#L86

and searching through the history there used to be a better explanation of why eslint ignores in this case

https://github.com/eslint/eslint/blob/361377fe6182510a5773438408f13f11a58232f9/lib/rules/prefer-const.js#L184

from that file
```
// - If `eslintUsed` is true, we cannot know where it was used from.
//   In this case, if the scope of the variable would change, it
//   skips the variable.
```

so I presume that is to stop this

```
if (true) {
    var a = 1;
}
return (<div>{a}</div>);
```
because in that case, converting the var to a const would change the meaning of the code (a would no longer be defined).

maybe we can only ignore eslintUsed variables if they are var so that `let` variables can still be warned?

**Comment by mysticatea** [comment]
@lukeapage Thank you for investigating.

> so I presume that is to stop this

`prefer-const` does not touch `var` declarations. :)
It's handling only `let` declarations.

Actually, the rule ignores variables if `eslintUsed` is `true`, because rules cannot distinguish reasons of used. (e.g., `/*exported*/` comments set `true` to `eslintUsed`)

**Comment by mysticatea** [comment]
Hmm, "used" is not modifications, so it may be no problem that we remove the check of `eslintUsed`.

**Comment by lukeapage** [comment]
How does the reasons of usage change whether prefer-const should be applied? Can you give an example?
I was going to follow up that eslintUsed could maybe instead be an array of scopes.. but do you have a better idea?

**Comment by mysticatea** [comment]
I'm sorry, I was confused. I found my memory.
Originally, `prefer-const` was warning the following patterns:

```js
let a;
for (let i = 0; i < 10; ++i) {
    a = doSomething();
    doSomething2(a);
}

// it can be modified as below.

for (let i = 0; i < 10; ++i) {
    let a = doSomething();
    doSomething2(a);
}
```

But in the following case (i.e., if there is read before write), the rule should not warn the code.

```js
let a;
for (let i = 0; i < 10; ++i) {
    doSomething2(a);  // this uses the value of the previous iteration.
    a = doSomething();
}
```

Now, in the following case, if variable `a` is "eslintUsed", the rule cannot check the location of the read.

```js
let a;
for (let i = 0; i < 10; ++i) {
    /* a code to set `eslintUsed`. */
    a = doSomething();
}
```

So, I wrote the condition `!(variable.eslintUsed && variable.scope !== writer.from)`.

**Comment by mysticatea** [comment]
But currently, `prefer-const` does not warn variables if the variable is initialized in the another scope from the declaration.

```js
let a;
for (let i = 0; i < 10; ++i) {
    a = doSomething(); // this is inside a different scope from `let a`, so this `a` is not warned.
}
```

So probably we can remove the check of `eslintUsed` simply.

**Comment by ljharb** [comment]
Just to confirm, `prefer-const` should however still *not* warn on:
```js
let foo;
const recursive = () => {
  return foo;
};
foo = {
  dontAsk: recursive
};
```

**Comment by mysticatea** [comment]
@ljharb It would be warned by default. It can be modified to:

```js
const foo = {
    dontAsk: () => {
        return foo
    }
}
```

Also, we can make the rule ignoring that case by [ignoreReadBeforeAssign](http://eslint.org/docs/rules/prefer-const#ignorereadbeforeassign) option.

**Comment by ljharb** [comment]
@mysticatea ah sorry. My mistake. https://github.com/eslint/eslint/issues/5822 Ended up being resolved by adding the option you mentioned.

**Comment by yannickcr** [comment]
> So probably we can remove the check of `eslintUsed` simply.

@mysticatea any news on this? This check is breaking `prefer-const` when used with `react/jsx-uses-vars` (https://github.com/yannickcr/eslint-plugin-react/issues/716).

**Comment by mysticatea** [comment]
Oh, I'm sorry.
I had lost the track of this issue.

I will fix this.

**Comment by mysticatea** [comment]
Oh, wait, but I'm not sure whether it's the issue of this original post or not.
`dynStyle` seems not the target of `react/jsx-uses-vars` rule.

Maybe should we separate issue?

**Comment by yannickcr** [comment]
The issue is that any variable that is marked as used by [`context.markVariableAsUsed(name)`](http://eslint.org/docs/developer-guide/working-with-rules#the-context-object) (which set `eslintUsed` to `true` on the variable) will be ignored by `prefer-const`.

It was the case for `dynStyle` that was marked as used by `react/jsx-uses-vars` (fixed now since ESLint seems to handle it correctly without any help https://github.com/yannickcr/eslint-plugin-react/commit/d4dee487b92aebe7377a38cb6493c44ebc3569b8).

But we'll still have the same problem with `JSXOpeningElement`:

```jsx
let App = <div />; // should warn "'App' is never reassigned. Use 'const' instead" but will not
<App />
```

So it is still the same issue, just on another node type.

But I can open another issue if it makes it easier for you.

**Comment by mysticatea** [comment]
Thank you for explaining.
I understood it.
